File: packages/opencv_object_detection/android/src/test/java/com/grozziie/opencv_object_detection/YoloPostprocessorTest.java
package com.grozziie.opencv_object_detection;

import org.junit.Test;

import java.util.List;

import static org.junit.Assert.assertEquals;

public class YoloPostprocessorTest {
    @Test
    public void decodesChannelsFirstAndRemovesPerson() {
        float[] output = new float[]{
                128f, 64f,
                128f, 64f,
                100f, 20f,
                100f, 20f,
                0.1f, 0.95f,
                0.90f, 0.1f,
        };

        List<DetectionModels.Detection> detections = YoloPostprocessor.decode(
                output,
                new int[]{1, 6, 2},
                List.of("person", "Bottle"),
                transform(256, 256, 1f, 0f, 0f),
                0.25f,
                0.50f,
                5
        );

        assertEquals(1, detections.size());
        assertEquals("Bottle", detections.get(0).label);
        assertEquals(1, detections.get(0).classIndex);
        assertEquals(0.90f, detections.get(0).confidence, 0.0001f);
    }

    @Test
    public void reversesLetterboxingIntoNormalizedSourceCoordinates() {
        float[] output = new float[]{128f, 128f, 128f, 128f, 0.90f};

        List<DetectionModels.Detection> detections = YoloPostprocessor.decode(
                output,
                new int[]{1, 5, 1},
                List.of("Bottle"),
                transform(100, 200, 1.28f, 64f, 0f),
                0.25f,
                0.50f,
                5
        );

        DetectionModels.Detection detection = detections.get(0);
        assertEquals(0f, detection.left, 0.0001f);
        assertEquals(0.25f, detection.top, 0.0001f);
        assertEquals(1f, detection.right, 0.0001f);
        assertEquals(0.75f, detection.bottom, 0.0001f);
    }

    @Test
    public void appliesClassAwareNmsAndResultLimit() {
        float[] output = new float[]{
                128f, 130f,
                128f, 130f,
                100f, 100f,
                100f, 100f,
                0.90f, 0.80f,
        };

        List<DetectionModels.Detection> detections = YoloPostprocessor.decode(
                output,
                new int[]{1, 5, 2},
                List.of("Bottle"),
                transform(256, 256, 1f, 0f, 0f),
                0.25f,
                0.50f,
                1
        );

        assertEquals(1, detections.size());
        assertEquals(0.90f, detections.get(0).confidence, 0.0001f);
    }

    @Test
    public void keepsOverlappingBoxesFromDifferentClasses() {
        float[] output = new float[]{
                128f, 128f,
                128f, 128f,
                100f, 100f,
                100f, 100f,
                0.90f, 0.10f,
                0.10f, 0.80f,
        };

        List<DetectionModels.Detection> detections = YoloPostprocessor.decode(
                output,
                new int[]{1, 6, 2},
                List.of("Bottle", "Chair"),
                transform(256, 256, 1f, 0f, 0f),
                0.25f,
                0.50f,
                5
        );

        assertEquals(2, detections.size());
        assertEquals("Bottle", detections.get(0).label);
        assertEquals("Chair", detections.get(1).label);
    }

    @Test
    public void removesCandidatesBelowConfidenceThreshold() {
        float[] output = new float[]{128f, 128f, 100f, 100f, 0.24f};

        List<DetectionModels.Detection> detections = YoloPostprocessor.decode(
                output,
                new int[]{1, 5, 1},
                List.of("Bottle"),
                transform(256, 256, 1f, 0f, 0f),
                0.25f,
                0.50f,
                5
        );

        assertEquals(0, detections.size());
    }

    private DetectionModels.LetterboxTransform transform(
            int sourceWidth,
            int sourceHeight,
            float scale,
            float padX,
            float padY
    ) {
        return new DetectionModels.LetterboxTransform(
                sourceWidth,
                sourceHeight,
                256,
                256,
                scale,
                padX,
                padY
        );
    }
}

File: packages/native_object_detection/android/src/test/kotlin/com/grozziie/native_object_detection/YoloPostprocessorTest.kt
package com.grozziie.native_object_detection

import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

class YoloPostprocessorTest {
    @Test
    fun `filters person and suppresses overlapping boxes`() {
        val anchors = 3
        val output = FloatArray(6 * anchors)
        setAnchor(output, anchors, 0, 320f, 320f, 300f, 300f, 0.95f, 0.10f)
        setAnchor(output, anchors, 1, 320f, 320f, 320f, 320f, 0.05f, 0.90f)
        setAnchor(output, anchors, 2, 325f, 325f, 320f, 320f, 0.05f, 0.80f)

        val detections = YoloPostprocessor.decode(
            output = output,
            outputDimensions = intArrayOf(1, 6, anchors),
            labels = listOf("person", "bottle"),
            transform = transform(sourceWidth = 640, sourceHeight = 640),
            confidenceThreshold = 0.60f,
            iouThreshold = 0.50f,
            maxResults = 5,
        )

        assertEquals(1, detections.size)
        assertEquals("bottle", detections.single().label)
        assertEquals(1, detections.single().classIndex)
        assertEquals(0.90f, detections.single().confidence, 0.0001f)
    }

    @Test
    fun `removes letterbox padding and returns normalized source boxes`() {
        val output = FloatArray(6)
        setAnchor(output, 1, 0, 320f, 320f, 320f, 160f, 0.05f, 0.90f)

        val detections = YoloPostprocessor.decode(
            output = output,
            outputDimensions = intArrayOf(1, 6, 1),
            labels = listOf("person", "bottle"),
            transform = transform(sourceWidth = 640, sourceHeight = 320),
            confidenceThreshold = 0.60f,
            iouThreshold = 0.50f,
            maxResults = 5,
        )

        val box = detections.single()
        assertEquals(0.25f, box.left, 0.0001f)
        assertEquals(0.25f, box.top, 0.0001f)
        assertEquals(0.75f, box.right, 0.0001f)
        assertEquals(0.75f, box.bottom, 0.0001f)
        assertTrue(box.right > box.left && box.bottom > box.top)
    }

    private fun setAnchor(
        output: FloatArray,
        anchorCount: Int,
        anchor: Int,
        centerX: Float,
        centerY: Float,
        width: Float,
        height: Float,
        personScore: Float,
        bottleScore: Float,
    ) {
        output[anchor] = centerX
        output[anchorCount + anchor] = centerY
        output[anchorCount * 2 + anchor] = width
        output[anchorCount * 3 + anchor] = height
        output[anchorCount * 4 + anchor] = personScore
        output[anchorCount * 5 + anchor] = bottleScore
    }

    private fun transform(sourceWidth: Int, sourceHeight: Int): LetterboxTransform {
        val scale = minOf(640f / sourceWidth, 640f / sourceHeight)
        return LetterboxTransform(
            sourceWidth = sourceWidth,
            sourceHeight = sourceHeight,
            modelWidth = 640,
            modelHeight = 640,
            scale = scale,
            padX = (640 - sourceWidth * scale) / 2f,
            padY = (640 - sourceHeight * scale) / 2f,
        )
    }
}
